### PR TITLE
Implement Trace Identifiers.

### PR DESCRIPTION
**Pull request by carlosalberto**
This is done to implement **Trace Identifiers**, as covered in the RFC:
 https://github.com/opentracing/specification/blob/master/rfc/trace_identifiers.md

Probably the most important question is whether we should have `traceId()` and `spanId()` instead of `tracerIdentifier()` and `spanIdentifier()` respectively - probably the important bit here is that it could (as mentioned in the past) break Tracer implementations 

(At this moment, doing this renaming would break `MockTracer`, as the current methods would have to be renamed to something like `longTraceId()` and `longSpanId()`. I'm definitely up for *actually* breaking it, but I wanted to get some feedback before ;) )

@yurishkuro @tedsuo @codingfabian @wu-sheng

### CODE REVIEW

**Comment by CodingFabian** [comment]
I would bikeshed for `traceId()` and `spanId()` because thats the signature of the methods we already implemented ;)

**Comment by yurishkuro** [comment]
I like the longer name, certainly less intrusive for existing implementations.

cc @pavolloffay @jpkrohling @objectiser

**Comment by jpkrohling** [comment]
> whether we should have traceId() and spanId() instead of tracerIdentifier() and spanIdentifier()

As a consumer of the API, I prefer `traceId()` instead of `tracerIdentifier()`, but I wonder why it's not following the standard `getTraceId()` convention.

> I'm definitely up for actually breaking it

+1, especially because we have not reached 1.0 yet.

**Comment by objectiser** [comment]
I would vote for `getTraceId()` and `getSpanId()` - and breaking is fine.

**Comment by hypnoce** [comment]
If we are ok to break things, I would go for `traceId()` and `spanId()` as it is the same naming conventions as the current version of the standard : https://w3c.github.io/distributed-tracing/report-trace-context.html#field-value
As for using the `get` prefix, this is more something esthetic. Both  I would suggest choose one option and stick to it throughout the API (my personnal choice here would be without the `get` prefix).

**Comment by objectiser** [comment]
@hypnoce get/set is a Java bean convention for properties - and is used in the current Java API, e.g. [getBaggageItem](https://github.com/opentracing/opentracing-java/blob/master/opentracing-api/src/main/java/io/opentracing/Span.java#L133), [setOperationName](https://github.com/opentracing/opentracing-java/blob/master/opentracing-api/src/main/java/io/opentracing/Span.java#L140), etc.

**Comment by hypnoce** [comment]
@objectiser Indeed it's a specification of Java Beans. They serve a specific purpose. Nevertheless, this convention is not consitently respected in the API https://github.com/opentracing/opentracing-java/blob/58846260a696f3cc06ae7c4ef1798d95a170300c/opentracing-api/src/main/java/io/opentracing/SpanContext.java#L35

From what I could see throughout the API, I beleive when there is a setter then the Java bean convention is used. If there is only a getter, the method has the same name as the property.

**Comment by carlosalberto** [comment]
@wu-sheng @adriancole @felixbarny Something to say on this? ;)

**Comment by carlosalberto** [comment]
Also pinging @tylerbenson as the author of DataDog's implementation ;)

**Comment by yurishkuro** [comment]
I would also vote to not use `get` prefix. And using the long form `Identifier` avoids unpleasant clashes with existing internal APIs.

**Comment by tylerbenson** [comment]
I forgot to mention that before... I actually prefer sticking with `Id`.  Long form is kind of grating to me.

**Comment by felixbarny** [comment]
I'm fine with either of the method names, but would prefer traceId() and spanId() to align with TraceContext. FWIW, the [Elastic APM OT bridge](https://github.com/elastic/apm-agent-java/blob/master/apm-opentracing/src/main/java/co/elastic/apm/opentracing/ApmSpan.java) does not currently offer ways to access the IDs so for us, it would not break anything.

Glad to see this change coming!

**Comment by carlosalberto** [comment]
Trying to re-take this discussion: it looks we are slightly more towards `traceId()` and `spanId()`.

Something to add anybody? Something to add @yurishkuro ? ;) Else, I will update the PR (including `MockTracer`) in the next days.

**Comment by carlosalberto** [comment]
Hey all,

Have updated the trace/span identifiers to match the described requirements. Let me know ;)

File: opentracing-api/src/main/java/io/opentracing/SpanContext.java [hunk 1]
```diff
@@ -26,6 +26,24 @@
  * @see Span#getBaggageItem(String)
  */
 public interface SpanContext {
+    /**
+     * Return the ID of the trace.
+     *
+     * Globally unique. Every span in a trace shares this ID.
+     *
+     * @return the trace ID for this context.
```
Comment (yurishkuro) [inline opentracing-api/src/main/java/io/opentracing/SpanContext.java]: what about returning empty string and never null?

Comment (carlosalberto) [reply opentracing-api/src/main/java/io/opentracing/SpanContext.java]: Oh, good point.

**Comment by hypnoce** [comment]
Hey @carlosalberto 
Thanks for the update.

> Trying to re-take this discussion: it looks we are slightly more towards `traceId()` and `spanId()`.

The `to` prefix is to avoid breaking the API ?

**Comment by tylerbenson** [comment]
@carlosalberto I'm confused... I thought we decided on `traceId()` and `spanId()`.  Where did we decide to add the `to` prefix?

**Comment by carlosalberto** [comment]
@tylerbenson Hey Tyler

Sorry for the PR without a previous explanation - so after discussing it on the Cross-Language call last Friday, based the fear of both breaking tracers *and* to signal users that retrieving such Ids _may_ incur in object allocation, we decided to go this way.

Of course, this is not set in stone, so we can have a few iterations and pondering on this item ;) Let us know.

**Comment by carlosalberto** [comment]
@hypnoce Yes, and also to signal potencial allocation when converting the ids in whatever format they have natively to `String` ;)

**Comment by carlosalberto** [comment]
After https://github.com/opentracing/specification/pull/125 I'm wondering if we could have this one merged (to the `v0.32.0` branch, that is, which is where experimental new features are supposed to land).

@yurishkuro @tylerbenson @objectiser @felixbarny @hypnoce

File: opentracing-api/src/main/java/io/opentracing/SpanContext.java [hunk 2]
```diff
@@ -26,6 +26,28 @@
  * @see Span#getBaggageItem(String)
  */
 public interface SpanContext {
+    /**
+     * Return the ID of the trace.
+     *
+     * Globally unique. Every span in a trace shares this ID.
+     *
+     * The empty string is a valid return value, while null is not.
```
Comment (objectiser) [inline opentracing-api/src/main/java/io/opentracing/SpanContext.java]: Do we need to clarify what an empty string represents? i.e. no active trace instance.
Also do we need other text that indicates the return value of toTraceId and toSpanId are either both empty strings (no active trace) or both have ids? For example, I don't think it would be valid to have an empty trace id but a value returned from span id?

Comment (carlosalberto) [reply opentracing-api/src/main/java/io/opentracing/SpanContext.java]: Oh, that's a good point. IMHO it would make sense to mention that the empty string *ought* to represent missing Trace Identifiers info (and, if this gets approved, make also the note about Span ID & Trace ID working together).

Sounds like a good plan?

cc @yurishkuro @tedsuo

Comment (yurishkuro) [reply opentracing-api/src/main/java/io/opentracing/SpanContext.java]: iirc the original idea was to allow empty string to tracers that are unable to provide a single trace id (e.g. "do not support them", or more likely "this span has many trace IDs").

Comment (tedsuo) [reply opentracing-api/src/main/java/io/opentracing/SpanContext.java]: I would recommend we keep the rules simple, and just say that either method may return an empty string, meaning that the method is not supported for that span. 

I doubt implementations will provide one but not the other, but there's no need to make it a rule.

**Comment by carlosalberto** [comment]
Hey all, 

Went ahead and clarified in the docs the meaning of an empty `String`, but without mention of the need of having *both* of them present or none (to keep things flexible).

Let me know if this is good to merge

cc @objectiser @yurishkuro @pavolloffay @tedsuo

File: opentracing-noop/src/main/java/io/opentracing/noop/NoopSpanBuilder.java [hunk 3]
```diff
@@ -80,6 +80,5 @@ public Span startManual() {
         return NoopSpanImpl.INSTANCE;
     }
 
-    @Override
```
Comment (pavolloffay) [inline opentracing-noop/src/main/java/io/opentracing/noop/NoopSpanBuilder.java]: why is this removed?

File: opentracing-mock/src/main/java/io/opentracing/mock/MockSpan.java [hunk 4]
```diff
@@ -202,9 +202,12 @@ public MockContext(long traceId, long spanId, Map<String, String> baggage) {
         }
 
         public String getBaggageItem(String key) { return this.baggage.get(key); }
+        public String toTraceId() { return String.valueOf(traceId); }
+        public String toSpanId() { return String.valueOf(spanId); }
         public long traceId() { return traceId; }
         public long spanId() { return spanId; }
 
+
```
Comment (objectiser) [inline opentracing-mock/src/main/java/io/opentracing/mock/MockSpan.java]: nit: should remove this line

File: opentracing-noop/src/main/java/io/opentracing/noop/NoopSpanBuilder.java (hunk 3 above)
Comment (carlosalberto) [reply opentracing-noop/src/main/java/io/opentracing/noop/NoopSpanBuilder.java]: Damn, my bad.

**Comment by carlosalberto** [comment]
@pavolloffay @objectiser Small fixes done ;)

**Comment by carlosalberto** [comment]
Hey all!

Shall we merge this PR?

**Comment by hypnoce** [comment]
Hi all,

I have a small question regarding the format of the strings returned : do they have to obey the w3c format (hex strings) ? Or it's implementation detail ?
Since the w3c is mentioned in the spec, I was wondering.

Thanks

**Comment by tedsuo** [comment]
@hypnoce there is no restriction. Basically, this should return the “ToString()” equivalent of whatever your tracer’s native ID’s are. 

I would like to add w3c specific methods later, but that spec is more complex than just ID’s and it’s not really adopted yet. These accessors are meant to work with any tracing system which supports OT.

**Comment by hypnoce** [comment]
@tedsuo thanks for the update

**Comment by carlosalberto** [comment]
Hey all,

Unless there's somebody against it, I will be merging this PR in the next days (it has already been approved by a pair of members of the community, and it has received proper tuning ;)

**Comment by carlosalberto** [comment]
@yurishkuro Done. Thanks again for the latest feedback. Let me know if this is good to go - else, let me know so I can tune docs a little bit more ;)

File: opentracing-api/src/main/java/io/opentracing/SpanContext.java [hunk 5]
```diff
@@ -26,6 +26,38 @@
...
+     * Return the ID of the trace.
+     *
+     * Should be globally unique. Every span in a trace shares this ID.
+     *
+     * An empty String will be returned if the tracer does not support this functionality
+     * (this is the case for no-op tracers, for example). null is an invalid return value.
+     *
+     * Observe there is no restriction regarding the format of the returned value, as this
+     * should simply be the equivalent of calling toString() on the native ID object
+     * used by the tracer.
```
Comment (yurishkuro) [inline opentracing-api/src/main/java/io/opentracing/SpanContext.java]: This wording is problematic. It says "should ... ToString" which in many cases is the wrong thing to do, e.g. for int64 IDs. If we want to make a recommendation, it would be that the string format SHOULD match the representation of the IDs in http headers.

The javadoc here should ideally be repeating word for word the specification.

Comment (carlosalberto) [reply opentracing-api/src/main/java/io/opentracing/SpanContext.java]: Oh, good point and good example. So I feel that we should remove this clarification then (if later we add something in the specification itself, we can add it here). Would that make sense to you?

Comment (yurishkuro) [reply opentracing-api/src/main/java/io/opentracing/SpanContext.java]: Yes, can remove it altogether

**Comment by yurishkuro** [comment]
🚢

**Comment by carlosalberto** [comment]
Merged :)
